lib/src/Uncertainty/Distribution/FrechetFactory.cxx

```diff
@@ -0,0 +1,130 @@
+//                                               -*- C++ -*-
+/**
+ *  @brief Factory for Frechet distribution
+ *
+ *  Copyright 2005-2017 Airbus-EDF-IMACS-Phimeca
+ *
+ *  This library is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#include "openturns/FrechetFactory.hxx"
+#include "openturns/GumbelFactory.hxx"
+#include "openturns/MaximumLikelihoodFactory.hxx"
+#include "openturns/SpecFunc.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(FrechetFactory)
+
+static const Factory<FrechetFactory> Factory_FrechetFactory;
+
+/* Default constructor */
+FrechetFactory::FrechetFactory()
+  : DistributionFactoryImplementation()
+{
+  // Nothing to do
+}
+
+/* Virtual constructor */
+FrechetFactory * FrechetFactory::clone() const
+{
+  return new FrechetFactory(*this);
+}
+
+/* Here is the interface that all derived class must implement */
+
+FrechetFactory::Implementation FrechetFactory::build(const Sample & sample) const
+{
+  return buildAsFrechet(sample).clone();
+}
+
+FrechetFactory::Implementation FrechetFactory::build(const Point & parameters) const
+{
+  return buildAsFrechet(parameters).clone();
+}
+
+FrechetFactory::Implementation FrechetFactory::build() const
+{
+  return buildAsFrechet().clone();
+}
+
+Frechet FrechetFactory::buildAsFrechet(const Sample & sample) const
+{
+  const Scalar size = sample.getSize();
+  if (size == 0) throw InvalidArgumentException(HERE) << "Error: cannot build a Frechet distribution from an empty sample";
+  if (sample.getDimension() != 1) throw InvalidArgumentException(HERE) << "Error: can build a Frechet distribution only from a sample of dimension 1, here dimension=" << sample.getDimension();
+  const Scalar xMin = sample.getMin()[0];
+  Scalar gamma = xMin - std::abs(xMin) / (2.0 + size);
+  if (!SpecFunc::IsNormal(gamma)) throw InvalidArgumentException(HERE) << "Error: cannot build a Frechet distribution if data contains NaN or Inf";
+  // Convert the translated sample in logarithmic scale, in order for the new sample to be distributed according to the Gumbel distribution
+  Sample logSample(size, 1);
+  for (UnsignedInteger i = 0; i < size; ++i)
+    logSample[i][0] = std::log(sample[i][0] - gamma);
```

Comment:
If all values are nonnegative, one being zero, gamma is zero and we compute log(0) here.